File: mobile/lib/services/auth_service.dart
import 'api_service.dart';

/// Service d'authentification — Login par téléphone + OTP
class AuthService {
  final ApiService _api;

  AuthService(this._api);

  /// Envoie un OTP au numéro de téléphone
  Future<Map<String, dynamic>> sendOTP(String phone) async {
    // MOCK: En attendant le backend
    await Future.delayed(const Duration(seconds: 1));
    return {'success': true, 'message': 'OTP envoyé à $phone'};

    // REAL:
    // final response = await _api.post('/auth/send-otp', data: {'phone': phone});
    // return response.data;
  }

  /// Vérifie le code OTP
  Future<Map<String, dynamic>> verifyOTP(String phone, String code) async {
    // MOCK
    await Future.delayed(const Duration(seconds: 1));
    if (code == '123456') {
      return {
        'success': true,
        'token': 'mock_jwt_token_${DateTime.now().millisecondsSinceEpoch}',
        'refreshToken': 'mock_refresh_token',
        'user': {
          'id': 'user-001',
          'phone': phone,
          'name': 'Privat',
          'role': null, // pas encore choisi
        },
      };
    }
    return {'success': false, 'message': 'Code invalide'};

    // REAL:
    // final response = await _api.post('/auth/verify-otp', data: {'phone': phone, 'code': code});
    // return response.data;
  }

  /// Définit le rôle de l'utilisateur
  Future<Map<String, dynamic>> setRole(String userId, String role) async {
    // MOCK
    await Future.delayed(const Duration(milliseconds: 500));
    return {'success': true, 'role': role};

    // REAL:
    // final response = await _api.put('/auth/role', data: {'userId': userId, 'role': role});
    // return response.data;
  }

  /// Rafraîchit le token JWT
  Future<String?> refreshToken(String refreshToken) async {
    // MOCK
    return 'new_mock_jwt_token';

    // REAL:
    // final response = await _api.post('/auth/refresh', data: {'refreshToken': refreshToken});
    // return response.data['token'];
  }
}

File: backend/services/auth-service/src/auth.service.ts
import { Injectable, UnauthorizedException, BadRequestException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { OtpService } from './otp.service';
import { v4 as uuidv4 } from 'uuid';

@Injectable()
export class AuthService {
  constructor(
    private readonly jwtService: JwtService,
    private readonly otpService: OtpService,
  ) {}

  /**
   * Envoie un OTP au numéro de téléphone
   */
  async sendOtp(phone: string) {
    // Valider le format du numéro (Côte d'Ivoire: +225 XX XX XX XX XX)
    const cleanPhone = phone.replace(/\s/g, '');
    
    // Générer OTP 6 chiffres
    const code = this.otpService.generateCode();
    
    // Stocker dans Redis avec expiration 5 min
    await this.otpService.storeCode(cleanPhone, code);
    
    // Envoyer par SMS via le Notification Service (RabbitMQ)
    // TODO: Publish to RabbitMQ queue 'sms.send'
    console.log(`📱 OTP for ${cleanPhone}: ${code}`); // Dev only

    return {
      success: true,
      message: 'Code envoyé par SMS',
      // En dev, on renvoie le code pour faciliter les tests
      ...(process.env.NODE_ENV === 'development' && { code }),
    };
  }

  /**
   * Vérifie l'OTP et crée/connecte l'utilisateur
   */
  async verifyOtp(phone: string, code: string) {
    const cleanPhone = phone.replace(/\s/g, '');
    
    // Vérifier le code
    const isValid = await this.otpService.verifyCode(cleanPhone, code);
    if (!isValid) {
      throw new UnauthorizedException('Code invalide ou expiré');
    }

    // Chercher ou créer l'utilisateur
    // TODO: Database query
    const user = {
      id: uuidv4(),
      phone: cleanPhone,
      name: null,
      role: null,
      created_at: new Date().toISOString(),
    };

    // Générer les tokens
    const payload = { sub: user.id, phone: user.phone, role: user.role };
    const token = this.jwtService.sign(payload);
    const refreshToken = this.jwtService.sign(payload, { expiresIn: '30d' });

    // Log dans audit_logs
    // TODO: Insert audit log

    return {
      success: true,
      token,
      refreshToken,
      user,
      isNewUser: true, // TODO: check if existing
    };
  }

  /**
   * Rafraîchit le JWT
   */
  async refreshToken(refreshToken: string) {
    try {
      const payload = this.jwtService.verify(refreshToken);
      const newToken = this.jwtService.sign({
        sub: payload.sub,
        phone: payload.phone,
        role: payload.role,
      });
      return { success: true, token: newToken };
    } catch {
      throw new UnauthorizedException('Refresh token invalide');
    }
  }

  /**
   * Définit le rôle de l'utilisateur
   */
  async setRole(userId: string, role: string) {
    const validRoles = ['acheteur', 'vendeur', 'livreur'];
    if (!validRoles.includes(role)) {
      throw new BadRequestException('Rôle invalide. Choisis: acheteur, vendeur, ou livreur');
    }

    // TODO: Update user in database
    // If livreur, mark as needing verification

    return {
      success: true,
      role,
      message: role === 'livreur'
        ? 'Rôle livreur choisi. Vérification d\'identité requise.'
        : `Rôle ${role} activé. Bienvenue sur OCALM !`,
    };
  }
}
